dlib/matrix/fft_size.h

```diff
@@ -0,0 +1,215 @@
+            initialiser list.
+
+            If default-constructed, the object is empty and in an invalid state.
+            That is, FFT functions will throw if attempted to be used with such 
+            an object.
+
+            If constructed with an initialiser list L, the object is properly
+            initialised provided:
+               - L.size() > 0 and L.size() <= 5
+               - L contains strictly positive values
+
+            Once the object is constructed, it is immutable.
+        !*/
+    private:
+        using container_type    = std::array<long,5>;
+        using const_reference   = container_type::const_reference;
+        using iterator          = container_type::iterator;
+        using const_iterator    = container_type::const_iterator;
+        
+        size_t _size = 0;
+        container_type _dims;
+        
+        iterator begin()
+        {
+            return _dims.begin();
+        }
+        
+        const_iterator begin() const
+        {
+            return _dims.begin();
+        }
+        
+        iterator end()
+        {
+            return _dims.begin() + _size;
+        }
+        
+        const_iterator end() const
+        {
+            return _dims.begin() + _size;
+        }
+
+    public:
+        fft_size() = default;
+        /*!
+            ensures
+                - this is properly initialised
+                - size() == 0
+                - is_valid() == false
+        !*/
+        
+        fft_size(std::initializer_list<long> dims)
+        {
+            DLIB_ASSERT(dims.size() > 0, "the initialiser list must be non-empty");
+            DLIB_ASSERT(dims.size() <= _dims.size(), "the initialiser list must have size less than 6");
+            std::copy(dims.begin(), dims.end(), _dims.begin());
+            _size = dims.size();
+            DLIB_ASSERT(is_valid(), "the initialiser list must contain strictly positive values");
+        }
+        /*!
+            requires
+                - dims.size() > 0 and dims.size() <= 5
+                - dims contains strictly positive values
+            ensures
+                - *this is properly initialised
+                - size() == dims.size()
+                - is_valid() == true 
+            throws
+                - dlib::fatal_error if requirements aren't satisfied.
+        !*/
+        
+        size_t size() const
+        {
+            return _size;
+        }
+        /*!
+            ensures
+                - returns the number of elements in *this
+        !*/
+
+        const_reference operator[](size_t index) const
+        {
+            DLIB_ASSERT(index < _size, "index " << index << " out of range [0," << _size << ")");
+            return _dims[index];
+        }
+        /*!
+            requires
+                - size() > 0
+                - index < size()
+            ensures
+                - returns a const reference to the element at position index
```

Comment:
Say dimension instead of element since otherwise it's ambiguous with respect to text that talks about fft sizes.  E.g. `dimprod` talks about number of elements too but it means something entirely different.

Reply:
yep